Refactor: Bug bounty program

maykelxyz opened the pull request:
# Refactor: Bug Bounty Program Schema

## Overview
Refactored the bug bounty program type definitions to provide granular, objective criteria for evaluating wallet security programs.

## Changes

### Replaced subjective categories with measurable attributes
**Before:** Used broad, subjective classifications (`COMPREHENSIVE`, `BASIC`, `DISCLOSURE_ONLY`, `NONE`)  
**After:** Introduced specific, verifiable fields that capture concrete program details

### New Type Definitions

**`BugBountyPlatform`** - Tracks which platform hosts the program (Self-hosted, HackerOne, Bugcrowd, etc.)

**`LegalProtectionType`** - Distinguishes between Safe Harbor clauses and other legal assurances for researchers

**`BugBountyProgramAvailability`** - Indicates current program status (`ACTIVE`, `INACTIVE`)

**`CoverageBreadth`** - Specifies what's covered: app, firmware, hardware, or full scope

**`BugBountyProgramSupport`** - Comprehensive program details including:
- Program start date
- Current availability status
- Coverage scope (with type-safe "at least one" coverage requirement)
- Reward ranges with currency
- Platform information
- Disclosure timeline
- Upgrade path availability
- Legal protections for researchers

## Benefits

- **Objective evaluation**: Enables consistent, measurable comparisons across wallets
- **Better transparency**: Makes security program details explicit and verifiable
- **Reference support**: Uses `WithRef` and `MustRef` types for proper documentation linkage

## Migration Notes

Existing wallets using the old enum-based system will need data migration to populate the new structured fields.

## Hardware Wallet Updates
### Wallet Updates for new structured fields
- Bitbox
- Cypherock
- Firefly
- Gridplus
- Keystone
- Ledger
- Ngrave
- Onekey
- Trezor
## Software Wallet Updates
### Wallet Updates for new structured fields
- Ambire
  - Updated Ambire's bug bounty program to be inactive. [Reference](https://blog.ambire.com/ambire-x-immunefy-bug-bounty-audit-our-code-and-earn-rewards/) 
- ImToken
- MtPelerin

Issue #230

maykelxyz committed:
[WIP] added objective and detailed info for bug bounty programs

maykelxyz committed:
Merge branch 'beta' into refactor-bug-bounty-program

maykelxyz committed:
Refactor: Update bug bounty

Aims to make bug bounty program objective

maykelxyz committed:
Refactor: Update bug bounty question to fit updated criteria

maykelxyz committed:
Feat: Add detailed info for new bug bounty details

maykelxyz committed:
Refactor: Remove BugbountyProgramType

BugBountyProgramType is replaced with a much more objective type

maykelxyz committed:
Chore: pnpm lint

maykelxyz committed:
fix: lint

maykelxyz committed:
Chore: Remove BugBountyProgramType

maykelxyz committed:
Refactor: Improve bug bounty rating

maykelxyz committed:
Refactor: Include platform and platformUrl

maykelxyz committed:
Feat: Add legal protection

maykelxyz committed:
Refactor: order of description

maykelxyz committed:
Feat: add legal protection to gridplus

polymutex commented on src/schema/features/security/bug-bounty-program.ts:
```diff
@@ -1,12 +1,67 @@
+	/**
+	 * URL or reference to the legal protection language
+	 */
+	reference?: string
```
Instead of URL fields, please use the `WithRef` wrapper type instead in general (check how it's used in other feature types). `WithRef` standardizes the way reference URLs are expressed across the wallet data and makes it easier to display them consistently in the UI later.

In this case however, if you implement `legalProtection: Support<LegalProtection>` as per my other comment suggestion, then any remaining value of `LegalProtection.type` must come with a URL. In such a case, you can use `MustRef` instead of `WithRef`, which forces specifying a URL (whereas it is optional with `WithRef`.

So something like this:

```typescript

export enum LegalProtectionType {
	SAFE_HARBOR = 'SAFE_HARBOR',
	LEGAL_ASSURANCE = 'LEGAL_ASSURANCE',
}

export type LegalProtection = MustRef<{
  type: LegalProtectionType
  // ... other fields here...
}>

// [...]

export interface BugBountyProgramSupport {
  // [... other fields...]

  legalProtections: Support<LegalProtection>
}
```

This will force either `legalProtections: unsupported`, or `legalProtections: supported({ ... })` and require a `ref` URL to be specified inside the `supported({ ... })` block.

polymutex commented on src/schema/features/security/bug-bounty-program.ts:
```diff
@@ -1,12 +1,67 @@
+	/**
+	 * Whether they participate in #legalbugbounty or similar standardization
+	 */
+	standardizedLanguage?: boolean
```
Is this referring to [this project](https://github.com/EdOverflow/legal-bug-bounty)? It doesn't look super standard. IMO we should remove this, and if we find a wallet with a bug bounty that uses this project we can add it at that time (possibly as a different `LegalProtectionType` enum value)

polymutex commented on src/schema/features/security/bug-bounty-program.ts:
```diff
@@ -1,12 +1,67 @@
+	FIRMWARE_ONLY = 'FIRMWARE_ONLY',
+	HARDWARE_ONLY = 'HARDWARE_ONLY',
 	NONE = 'NONE',
 }
```
As with other feature data, I recommend keeping this set small to the set of options we're sure are needed, in order not to make the attribute logic more complicated than it needs to be. It's easy to add new feature data later, harder to remove them from consideration.

In this case for example, the `NONE` option seems useless: what kind of bug bounty program covers nothing at all?
It's also not clear what the difference is between "PARTIAL" and "APP_ONLY" (aren't "APP_ONLY"/"FIRMWARE_ONLY"/"HARDWARE_ONLY" all by definition also partial?).

IMO this enum can either be deleted (and the implicit value is that everything is covered by the bug bounty program), or could be replaced by something like `coverageBreadth: 'ALL_VARIANTS' | AtLeastOneVariant`, which allows selecting by Variant instead. Look at the `SecurityAudit.variantsScope` field in `src/schema/features/security/security-audits.ts` as an example.

maykelxyz replied:
This is noted, and makes more sense.

> Look at the `SecurityAudit.variantsScope` field in `src/schema/features/security/security-audits.ts` as an example.

Checked `src/schema/features/security/security-audits.ts`, and will follow this format. 

```ts
	coverageBreadth: AtleastOneScope | 'FULL_SCOPE'
```

is in WIP

maykelxyz replied:
```ts
/**
 * A set of at least one coverage breadth
 */
export type AtLeastOneCoverageBreadth = NonEmptySet<CoverageBreadth>


	/**
	 * The coverage breadth of the bug bounty program
	 */
	coverageBreadth: AtLeastOneCoverageBreadth | 'FULL_SCOPE'
```

Improve coverageBreadth feature to the code above

polymutex commented on src/schema/features/security/bug-bounty-program.ts:
```diff
@@ -15,24 +70,65 @@ export enum BugBountyProgramType {
-	 * The type of bug bounty program implemented
+	 * The date the bug bounty program started
+	 */
+	dateStarted?: Date
```
All feature data in general must not be `undefined` (therefore, field names should never be followed by `?`). This is a general style we've followed (there's actually a test for it at `deploy/misc-tests/no_undefined_in_wallet_features.sh` though it only actually checks `src/schema/features.ts` so it didn't catch this). The rationale is that allowing `undefined` makes it difficult to look at a wallet data file and figure out what's missing and still needs to be filled out. So every field is required. For fields where we don't have any data yet, we can put `null`. This way, searching for `null` in a wallet file gives a comprehensive view of all the fields we still need data for.

(Yes I know there's an exception to this in `src/schema/features/security/security-audits.ts` right now; maybe for another PR to clean that up...)

Anyway, I don't think this field should be `dateStarted: Date | null` either. IMO, data entry for the bug bounty program should be all-or-nothing for simplicity. In other words, either we have full feature data for a wallet's `BugBountyProgramSupport`, or we have nothing. The `VariantFeature` wrapper type actually already takes care of this (look at `WalletBaseFeatures.security.bugBountyProgram`) as it already makes it `null`-able.

So for the purpose of `BugBountyProgramSupport`, every field should be non-null, non-undefined. I think that should keep things simpler to handle in the attribute evaluation logic.

Also, specifically for `dateStarted`, please use the `CalendarDate` type (from `src/types/date.ts`) as it is easier to input than JavaScript dates.

maykelxyz replied:
Resolved

polymutex commented on src/schema/features/security/bug-bounty-program.ts:
```diff
@@ -15,24 +70,65 @@ export enum BugBountyProgramType {
+	 * The minimum reward for a bug bounty program
 	 */
-	type: BugBountyProgramType
+	minimumReward?: number
```
I suggest replacing this with:

```typescript
rewards: Support<{
     minimum: number
     maximum: number
     currency: string
   }>
```

The `Support` wrapper type is useful for data that only makes sense to populate in case that a wallet supports something (in this case, if its bug bounty program supports rewards). This way it can be written like this in wallet data files:

```typescript
rewards: supported({
  minimum: 1000,
  maximum: 10000,
  currency: 'USD',
})
```

or:

```typescript
rewards: notSupported
```

polymutex commented on src/schema/features/security/bug-bounty-program.ts:
```diff
@@ -15,24 +70,65 @@ export enum BugBountyProgramType {
 	/**
 	 * URL to the bug bounty program details
 	 */
 	url?: string
```
Remove field and wrap in `WithRef` instead

polymutex commented on src/schema/features/security/bug-bounty-program.ts:
```diff
@@ -15,24 +70,65 @@ export enum BugBountyProgramType {
+	/**
+	 * The URL to the bug bounty program
+	 */
+	platformUrl?: string
```
I think this is probably going to be the same as the other URL already part of this type, so no need for another one here.

polymutex commented on src/schema/features/security/bug-bounty-program.ts:
```diff
@@ -15,24 +70,65 @@ export enum BugBountyProgramType {
 	/**
 	 * Additional details about the bug bounty program implementation
 	 */
 	details?: string
```
No longer needed; the new fields being added in this PR should remove the need to have a freeform `details` field.

maykelxyz replied:
Resolved

polymutex commented on src/schema/features/security/bug-bounty-program.ts:
```diff
@@ -15,24 +70,65 @@ export enum BugBountyProgramType {
+	 * The disclosure process of the bug bounty program;
+	 * number of days, privacy of reports, etc.
+	 */
+	disclosureProcess?: string
```
The goal of this PR is to make this not freeform, so let's either remove this, or keep it but structured, like:

```typescript
disclosure: Support<{ numberOfDays: number }>
```

polymutex commented on src/schema/features/security/bug-bounty-program.ts:
```diff
@@ -1,12 +1,67 @@
+ * Types of legal protection provided to security researchers
+ */
+export enum LegalProtectionType {
+	SAFE_HARBOR = 'SAFE_HARBOR', // Explicit Safe Harbor language
+	LEGAL_ASSURANCE = 'LEGAL_ASSURANCE', // Pledges not to sue, similar protections
+	NONE = 'NONE', // No legal protections mentioned
+}
```
Please put these comments as `/** JSDoc comments */` above each enum value instead, so that they show up in type documentation tooltips in IDEs.

maykelxyz replied:
Resolved

polymutex commented on src/schema/features/security/bug-bounty-program.ts:
```diff
@@ -15,24 +70,65 @@ export enum BugBountyProgramType {
+	/**
+	 * The legal protections offered to security researchers
+	 */
+	legalProtections?: LegalProtection
```
`legalProtections: Support<LegalProtection>`

maykelxyz replied:
Resolved

polymutex commented on src/schema/features/security/bug-bounty-program.ts:
```diff
@@ -1,12 +1,67 @@
+export enum LegalProtectionType {
+	SAFE_HARBOR = 'SAFE_HARBOR', // Explicit Safe Harbor language
+	LEGAL_ASSURANCE = 'LEGAL_ASSURANCE', // Pledges not to sue, similar protections
+	NONE = 'NONE', // No legal protections mentioned
```
Can remove NONE here if using `legalProtection: Support<LegalProtection>` as per my other comment suggestion.

maykelxyz replied:
Resolved

polymutex reviewed (changes requested):
Great direction so far. I only reviewed the feature data (not the attribute logic yet). Overall this is the right approach, just some stylistic/type changes to be more in line with the way other wallet feature data is usually expressed.

maykelxyz committed:
Merge branch 'beta' into refactor-bug-bounty-program

maykelxyz committed:
Refactor: update to be in line with wallet feat data

polymutex commented on src/schema/features/security/bug-bounty-program.ts:
```diff
@@ -1,39 +1,114 @@
+ * The availability of the bug bounty program
+ */
+export enum BugBountyProgramAvailability {
+	ACTIVE = 'ACTIVE', // Running now, accepting reports
+	INACTIVE = 'INACTIVE', // Temporarily paused
+	NEVER = 'NEVER', // Never existed
+}
```
Put comments as JSDoc

maykelxyz replied:
Resolved

polymutex commented on src/schema/features/security/bug-bounty-program.ts:
```diff
@@ -1,39 +1,114 @@
+	 * The date the bug bounty program started
 	 */
-	type: BugBountyProgramType
+	dateStarted: Date
```
Use CalendarDate rather than Date, it's easier to enter.

maykelxyz replied:
Resolved

polymutex commented on src/schema/features/security/bug-bounty-program.ts:
```diff
@@ -1,39 +1,114 @@
+	 * The disclosure process of the bug bounty program;
+	 * number of days, privacy of reports, etc.
+	 */
+	disclouse: Support<{
```
Typo: `disclosure`

maykelxyz replied:
Resolved

polymutex commented on src/schema/features/security/bug-bounty-program.ts:
```diff
@@ -1,39 +1,114 @@
+	/**
+	 * The platform of the bug bounty program
+	 */
+	platform?: BugBountyPlatform
```
Remove `?`

maykelxyz replied:
Resolved

maykelxyz committed:
fix: typo

maykelxyz committed:
refactor: updates for inline wallet data

polymutex commented on src/schema/features/security/bug-bounty-program.ts:
```diff
@@ -1,39 +1,118 @@
+export enum BugBountyProgramAvailability {
+	ACTIVE = 'ACTIVE',
+	INACTIVE = 'INACTIVE',
+	NEVER = 'NEVER', 
```
So, I just realized that `BugBountyProgramAvailability.NEVER` basically means there is no bug bounty program. The other fields in `BugBountyProgramSupport` are required, but make no sense when there is no bug bounty program.

So instead, I suggest doing the following:

- Remove `NEVER` from `BugBountyProgramAvailability` (the only two values are `ACTIVE` and `INACTIVE`)
- Wrap the entire `BugBountyProgramSupport` type with `Support` (i.e. in `features.ts`, make the `security.bugBountyProgram` field be `bugBountyProgram: VariantFeature<Support<BugBountyProgramImplementation>>`)

This way, if a wallet supports a bounty program, it'll be written as:

```typescript
bugBountyProgram: supported<BugBountyProgramImplementation>({
  availability: BugBountyProgramAvailability.ACTIVE,
  // ... all the other BugBountyProgramImplementation fields...
})
```

And if it does not:

```typescript
bugBountyProgram: unsupported
```

And if we don't know:

```typescript
bugBountyProgram: null
```

No redundant fields this way.

polymutex commented on src/schema/attributes/security/bug-bounty-program.ts:
```diff
@@ -47,98 +87,120 @@ function noBugBountyProgram(): Evaluation<BugBountyProgramValue> {
-		},
-		details: markdown(`
-			{{WALLET_NAME}} implements a comprehensive bug bounty program that offers strong incentives for security researchers to find and report vulnerabilities. The program has a wide scope, competitive rewards, and a responsive disclosure process.
+function determineRating(support: BugBountyProgramSupport): Evaluation<BugBountyProgramValue> {
```
Attribute logic looks nice in general. Needs some rework to account for the refactored `BugBountyProgramImplementation` type since this PR started. When dealing with `Support<...>` fields, please use the `isSupported` type predicate which will let you access fields within the `<...>` part in a type-safe way (otherwise the type-checker will complain). Check `src/schema/attributes/security/scam-prevention.ts` for an example attribute that uses `isSupported` a bunch.
So for example you'd probably have

```typescript
if (!isSupported(bugBountyProgram)) { // Check for `bugBountyProgram: unsupported`
  return noBugBountyProgram()
}

// You can now access the fields of `bugBountyProgram` safely in the rest of this function, because the type-checker knows the `bugBountyProgram` is `supported` since the `if` above didn't match.
```

polymutex commented on src/schema/attributes/security/bug-bounty-program.ts:
```diff
@@ -196,36 +261,31 @@ export const bugBountyProgram: Attribute<BugBountyProgramValue> = {
 				mdParagraph(`
-					The hardware wallet implements a comprehensive bug bounty program with clear incentives and responsive processes.
-					It offers competitive rewards based on severity, has a transparent disclosure process, and provides upgrade paths for users.
+					The hardware wallet has an active bug bounty program with competitive rewards (e.g., up to $50,000),
```
It's fair to use "does it have non-zero rewards" as a criterion, but for now let's not write specific numbers here unless we actually use these exact thresholds in the evaluation criteria (which I don't think we should just yet)

maykelxyz replied:
REsolved

polymutex commented on src/schema/attributes/security/bug-bounty-program.ts:
```diff
@@ -196,36 +261,31 @@ export const bugBountyProgram: Attribute<BugBountyProgramValue> = {
 				mdParagraph(`
-					The hardware wallet implements a basic bug bounty program with limited scope or rewards.
-					However, it does provide a clear upgrade path for users when security issues are discovered.
+					The hardware wallet has a bug bounty program with rewards (e.g., up to $5,000),
```
Same, remove number

maykelxyz replied:
Resolved

polymutex commented on src/schema/attributes/security/bug-bounty-program.ts:
```diff
@@ -47,98 +87,120 @@ function noBugBountyProgram(): Evaluation<BugBountyProgramValue> {
+		? ` [View legal protection policy](${legalProtection?.reference})`
+		: ''
+
+	return `Legal Protection: The program provides ${protectionType} protections for security researchers conducting good faith security research.${standardization}${referenceLink}`
```
Use either the `refs` or the `popRefs` helper functions to exact `ref`s out of the feature data, and return them as the `references` field of `Evaluation`. This will cause them to be automatically displayed in the wallet page. Check other attributes for examples of these functions.

polymutex commented on src/schema/attributes/security/bug-bounty-program.ts:
```diff
@@ -47,98 +87,120 @@ function noBugBountyProgram(): Evaluation<BugBountyProgramValue> {
+		: ''
+
+	const hasRewards =
+		support.minimumReward != null &&
+		support.maximumReward != null &&
+		support.minimumReward !== 0 &&
+		support.maximumReward !== 0
+	const hasFullCoverage = support.coverageBreadth === CoverageBreadth.FULL
+	const hasLegalProtection =
+		support.legalProtections?.type !== LegalProtectionType.NONE &&
+		support.legalProtections?.type != null
+	const isActive = support.availability === BugBountyProgramAvailability.ACTIVE
+
+	const passesAll = isActive && hasFullCoverage && hasRewards && hasLegalProtection
+
+	const rating = passesAll
+		? Rating.PASS
+		: isActive || hasRewards || hasFullCoverage || hasLegalProtection
+			? Rating.PARTIAL
+			: Rating.FAIL
```
Seems like a reasonable approach, I like it!

polymutex commented on src/schema/attributes/security/bug-bounty-program.ts:
```diff
@@ -20,11 +24,46 @@ import { exempt, pickWorstRating, unrated } from '../common'
+		return `with rewards up to $${support.maximumReward.toLocaleString()}`
+	} else if (support.minimumReward) {
+		return `with rewards starting at $${support.minimumReward.toLocaleString()}`
+	}
```
Add a case here where `minimumReward` and `maximumReward` are both set to the same number (i.e. fixed reward size). Under the current code that would say something like "with rewards ranging from $1000 to $1000", but it should say something like "with a $1000 reward" in the case they are equal.

maykelxyz replied:
WIP

maykelxyz replied:
Resolved

polymutex reviewed (commented):
Great progress, I started looking at the attribute implementation in the comments below.

maykelxyz committed:
Refactor: Add a case for same min and max

maykelxyz committed:
[WIP]

polymutex commented on src/schema/attributes/security/bug-bounty-program.ts:
```diff
@@ -20,11 +24,53 @@ import { exempt, pickWorstRating, unrated } from '../common'
+		case CoverageBreadth.NONE:
+			return 'The program has no defined coverage scope.'
+		default:
+			return ''
+	}
+}
+
+function getRewardDescription(support: BugBountyProgramSupport): string {
+	const min = support.rewards?.minimum
+	const max = support.rewards?.maximum
+
+	if (typeof min === 'number' && typeof max === 'number') {
+		if (min === max) {
+			return `with a $${min.toLocaleString()} reward`
+		} else {
+			return `with rewards ranging from $${min.toLocaleString()} to $${max.toLocaleString()}`
+		}
+	} else if (typeof max === 'number') {
+		return `with rewards up to $${max.toLocaleString()}`
+	} else if (typeof min === 'number') {
+		return `with rewards starting at $${min.toLocaleString()}`
+	}
+
+	return ''
+}
```
Here's how you can do this more simply and without manual `typeof` checks:

```typescript
function getRewardDescription(support: BugBountyProgramSupport): string {
  if (!isSupported(support.rewards)) {
    return ''
  }
  const min = support.rewards.min  // No `?` operator required, clean `number` type already inferred
  const max = support.rewards.max
  // ...
}
```

In general, if you find needing to use `typeof` or `foo?.bar` anytime in attribute code, there's probably a cleaner way to do it :)

maykelxyz replied:
Refactored into:

```ts
function getRewardDescription(support: BugBountyProgramSupport): string {
	if (!isSupported(support.rewards)) return ''

	const { minimum: min, maximum: max } = support.rewards

	if (min != null && max != null) {
		return min === max
			? `with a $${min.toLocaleString()} reward`
			: `with rewards ranging from $${min.toLocaleString()} to $${max.toLocaleString()}`
	}

	if (max != null) return `with rewards up to $${max.toLocaleString()}`
	if (min != null) return `with rewards starting at $${min.toLocaleString()}`
	return ''
}
```

Null checks are for rewards if maximum is given and minimum is not. Please feel free to let me know any feedbacks

polymutex replied:
Sounds good!

maykelxyz committed:
Refactor: wrap bug bounty in support

maykelxyz committed:
Refactor: Update getRewardsDescription

maykelxyz committed:
Refactor: Improve coverage breadth

polymutex commented on src/schema/attributes/security/bug-bounty-program.ts:
```diff
@@ -7,24 +7,76 @@ import {
+	upgradePathAvailable?: boolean
+	platform?: BugBountyPlatform
+	platformUrl?: string
+	legalProtections?: LegalProtection
```
These fields are not read anywhere, so you can remove them.
If you do, you'll need to replace the code below that says

```typescript
unrated(bugBountyProgram, brand, {
				availability: undefined,	
				coverageBreadth: CoverageBreadth.NONE,
				upgradePathAvailable: false,
			})
```

with just

```typescript
unrated(bugBountyProgram, brand, null)
```

polymutex commented on src/schema/attributes/security/bug-bounty-program.ts:
```diff
@@ -47,98 +100,108 @@ function noBugBountyProgram(): Evaluation<BugBountyProgramValue> {
+	const hasFullCoverage = support.coverageBreadth === CoverageBreadth.FULL
+	const hasLegalProtection =
+		isSupported(support.legalProtections) &&
+		support.legalProtections?.type != null
```
Checking `support.legalProtections?.type != null` is no longer needed, since `legalProtections.type` is guaranteed to be non-null by the type system in the supported case now.

polymutex commented on src/schema/attributes/security/bug-bounty-program.ts:
```diff
@@ -47,98 +100,108 @@ function noBugBountyProgram(): Evaluation<BugBountyProgramValue> {
-			id: 'disclosure_only_program',
-			rating: Rating.PARTIAL,
-			displayName: 'Basic disclosure policy',
+			id: 'bug_bounty_available',
```
Please make this ID string depend on `isActive`, similar to how it's done for `displayName` below. The ID isn't currently used but is intended to be roughly unique for each type of rating the function can return.

polymutex commented on src/schema/attributes/security/bug-bounty-program.ts:
```diff
@@ -47,98 +100,108 @@ function noBugBountyProgram(): Evaluation<BugBountyProgramValue> {
+			coverageBreadth: support.coverageBreadth || CoverageBreadth.NONE,
 			upgradePathAvailable: support.upgradePathAvailable,
+			platform: support.platform,
+			legalProtections: isSupported(support.legalProtections) ? support.legalProtections : undefined,
```
Can remove all these fields if removing them from `BugBountyProgramValue`.

polymutex commented on src/schema/attributes/security/bug-bounty-program.ts:
```diff
@@ -47,98 +100,108 @@ function noBugBountyProgram(): Evaluation<BugBountyProgramValue> {
 					: 'Unfortunately, the wallet does not provide a clear upgrade path for users when security issues are identified.'
 			}
+			
+			${support.platform ? `For more information, visit their [bug bounty program page](${support.platform}).` : ''}
```
I don't think `support.platform` is a valid URL, since it's an enum value.
I think you can simply remove this whole line. The `ref`s will be displayed automatically below on the wallet page, so the evaluation description doesn't need to include links.

polymutex commented on src/schema/attributes/security/bug-bounty-program.ts:
```diff
@@ -47,98 +100,108 @@ function noBugBountyProgram(): Evaluation<BugBountyProgramValue> {
+			
+			${support.platform ? `For more information, visit their [bug bounty program page](${support.platform}).` : ''}
 		`),
 		howToImprove: markdown(`
```
`howToImprove` should be unset if the evaluation is passing, since there's nothing to improve.

So you can do `howToImprove: passesAll ? undefined : markdown(/* ... */)`, and remove the `${passesAll ? '- Continue maintaining and improving the bug bounty program’s transparency and responsiveness' : ''}` part below.

polymutex commented on src/schema/attributes/security/bug-bounty-program.ts:
```diff
@@ -47,98 +100,108 @@ function noBugBountyProgram(): Evaluation<BugBountyProgramValue> {
-			- Improve response times and transparency in the vulnerability handling process
-		`),
-	}
+	return `Legal Protection: The program provides ${protectionType} protections for security researchers conducting good faith security research.`
```
Maybe add `**` (bold) around the words "Legal Protection".

polymutex commented on src/schema/attributes/security/bug-bounty-program.ts:
```diff
@@ -47,98 +100,108 @@ function noBugBountyProgram(): Evaluation<BugBountyProgramValue> {
-			`),
+	// No bug bounty program at all
+	if (!hasBugBounty) {
+		return noBugBountyProgram()
```
I believe this is no longer possible to be reached. The function that calls `determineRating` already checks that `features.security.bugBountyProgram` is supported, and the type system ensures that `support.availability` is never `undefined`.

So I think you can delete the `determineRating` function, and replace its call below with a direct call to `bugBountyAvailable`.

polymutex reviewed (commented):
Almost there! I think this is good to go once the rest of the comments here are addressed.

maykelxyz committed:
Fix: no mention of specific monetary rewards

maykelxyz committed:
Refactor: Remove platform enum description

maykelxyz committed:
Refactor: how to improve will be empty if score is passing

maykelxyz committed:
Refactor: make legal protection bold

maykelxyz committed:
fix: id should vary based on activeness

maykelxyz committed:
Fix: Remove null check of legal protection

maykelxyz committed:
Refactor: remove unused vars

maykelxyz committed:
Refactor: Remove detemineRating and call bugBountyAvailable directly

maykelxyz committed:
Refactor: Update coverageBreadth

Use `	coverageBreadth: AtLeastOneCoverageBreadth | 'FULL_SCOPE'
`

maykelxyz committed:
Refactor: Improve coverage description using new coverage breadth

polymutex commented on src/schema/attributes/security/bug-bounty-program.ts:
```diff
@@ -7,24 +7,74 @@ import {
+	if (descriptions.length === 0) {
+		return ''
+	}
+	return `The program covers ${descriptions.join(', ')}.`
```
For two items, this would return "The program covers only the application layer, only firmware vulnerabilities.".

I think you can move the "only" to line 52 and out of the individual cases of the `switch`, so that it says "The program covers only the application layer, firmware vulnerabilities.".

Then you can use `commaListFormat` from `src/types/utils/text.ts` instead of just `join`, which handles the case of using "and" vs commas depending on how many items there are. `The program covers only ${commaListFormat(descriptions)}.` would then return "The program covers only the application layer and firmware vulnerabilities."

polymutex replied:
Unresolving this conversation; still need to move the word `only` to the last `return` statement so that it's not repeated in case there's more than one item.

polymutex commented on src/schema/attributes/security/bug-bounty-program.ts:
```diff
@@ -247,41 +311,23 @@ export const bugBountyProgram: Attribute<BugBountyProgramValue> = {
 
 		const { withoutRefs, refs } = popRefs<BugBountyProgramSupport>(
 			features.security.bugBountyProgram,
 		)
```
Import the `refs` function from `src/schema/reference.ts`, then change this to:

```typescript
const allRefs = mergeRefs(
  refs(features.security.bugBountyProgram),
  isSupported(features.security.bugBountyProgram.legalProtections) ? refs(features.security.bugBountyProgram.legalProtections) : undefined,
)
```

This way you can catch both references (the one in `features.security.bugBountyProgram` and the one in `features.security.bugBountyProgram.legalProtections`).

Then on line 335 you can replace `...(refs.length > 0 && { references: refs }),` with `...(allRefs.length > 0 && { references: allRefs }),`

maykelxyz committed:
Refactor: Use commaListFormat

maykelxyz committed:
Refactor: Include refs with legalProtections

maykelxyz committed:
Fix: Import required references

maykelxyz committed:
Fix: revert withoutRefs

maykelxyz committed:
Fix: coverage description

maykelxyz committed:
Feat: Update bitbox bug bounty

maykelxyz committed:
Feat: Update cypherock's bug bounty program

maykelxyz committed:
Fix: Cypherock's bug bounty program

maykelxyz committed:
Chore: Create sample bug bounty program for unrated

maykelxyz committed:
Feat: Add gridplus bug bounty references

maykelxyz committed:
Feat: Add keystone's bug bounty program

maykelxyz committed:
Feat: Add Ngrave's bug bounty program

maykelxyz committed:
Chore: Add bug rap in bug bounty platform

maykelxyz committed:
Feat: Add one key's bug bounty information

maykelxyz committed:
Feat: Update ledger's bug bounty program

maykelxyz committed:
Feat: Add trezor's bug bounty feature

maykelxyz committed:
Fix: Ledger and trezor are interchanged

maykelxyz committed:
Fix: Update Ambire's wallet to not support bug bounty

maykelxyz committed:
Fix: Update imtoken's bug bounty

maykelxyz committed:
Chore: Add Immunefi in bug bounty platform

maykelxyz committed:
Fix: Update Mt Pelerin's bug bounty program

maykelxyz committed:
Feat: Add imtoken reference

maykelxyz committed:
Fix: fix bug bounty dates

Using archive.org to check earliest date published and / or publish dates of blogs

maykelxyz committed:
Chore: pnpm lint --fix

maykelxyz committed:
Chore: pnpm lint --fix

maykelxyz committed:
Fix: update firefly's bug bounty program not supported

maykelxyz committed:
Fix: Trezor's legal protection explanation

maykelxyz committed:
Fix: Grammar checks

maykelxyz committed:
Fix: pnpm lint --fix

maykelxyz committed:
Merge branch 'beta' into refactor-bug-bounty-program

maykelxyz committed:
Chore: Add Bugcrowd and Intigriti to cspell.json

maykelxyz committed:
Refactor: Update bug bounty's platform to remove links
